### DISCUSSION TITLE
How does Theia listen to install and uninstall vscode plug-ins at runtime

### DISCUSSION THREAD

**Discussion by zx616456607** [eclipse-theia/theia#13194] [General] [answered]
Please give a code example, thank you

**Comment by martin-fleck-at** [comment]
Hi @zx616456607,

You can use the `HostedPluginSupport` to see if plugins were installed or uninstalled by listening to the `onDidChangePlugins` event. The event itself will not contain any information about what exactly was installed or uninstalled so you need to track that yourself by retrieving the up-to-date set of plugins using `plugins` from the same class and doing a diff.

You can see an example for querying the data in the `PluginsWidget`:
https://github.com/eclipse-theia/theia/blob/c97a49648e5fbc7becfd5042ff9155953329904b/packages/plugin-ext/src/main/browser/plugin-ext-widget.tsx#L32-L81

Another example where we have a keep the list of installed VS Code plugins is `VSXExtensionsModel`:
https://github.com/eclipse-theia/theia/blob/c97a49648e5fbc7becfd5042ff9155953329904b/packages/vsx-registry/src/browser/vsx-extensions-model.ts#L242-L268

I hope that helps you get started with this.

**Reply by zx616456607** [reply]
version: 1.41.0
The first question: after uninstall plug-ins, 'this. puginSupport.plugins' access to or the old plugin list, there is no way to calculate the current which the plug-in uninstall.
Second question: you have installed the plug-in, different versions of the same plug-in installation, 'this.pluginSupport.plugins' unable to identify the latest version number, is there any other solutions?

**Reply by martin-fleck-at** [reply, marked as answer]
Hi @zx616456607,
You are right. I saw that there is an `outOfSync` flag on the plugin data which indicates that the installation status has changed. So I think if we can calculate the newly installed plugins and then filter the others through that flag we might get the behavior that we want:
```
        let pluginsById = new Map(this.pluginService.plugins.map(plugin => [plugin.model.id + '_' + plugin.model.version, plugin]));
        this.pluginService.onDidChangePlugins(() => {
            const newPluginsById = new Map(this.pluginService.plugins.map(plugin => [plugin.model.id + '_' + plugin.model.version, plugin]));
            const installed = [...newPluginsById.keys()].filter(newPluginId => !pluginsById.has(newPluginId)).map(id => newPluginsById.get(id)!);
            const uninstalled = [...newPluginsById.values()].filter(plugin => !installed.includes(plugin) && plugin.outOfSync);
            console.log('INSTALLED', installed);
            console.log('UNINSTALLED', uninstalled);
            pluginsById = newPluginsById;
        });
```

Regarding your second question: It seems that there is a bug somewhere indeed. I tested it when having a version 0.0.1 installed and then installing a version 0.0.2, I still get 0.0.1 reported as deployed plugin. I'll investigate this a bit more later today.

**Reply by zx616456607** [reply]
HostedPluginSupport onDidChangePluginsEmitter in class can return to the latest list of plugins? undefined is currently returned

**Reply by martin-fleck-at** [reply]
So I did some digging into the code and it seems that plugins are handled like that on purpose:

https://github.com/eclipse-theia/theia/blob/118e514a8073166cd3d6293a47c0a4ce14831e3b/packages/plugin-ext/src/hosted/node/plugin-service.ts#L111-L134

Specifically, the the plugin system was modified to handle uninstallation and the presence of multiple versions in this [PR](https://github.com/eclipse-theia/theia/commit/bfa1b44980421aa50a55d36ff5bff40e517e39ec) so that:
 - Most references to plugins include the version number through typing in most cases
 - Uninstalling a plugin on the frontend will not delete the deployed files, and the plugin will remain active on the frontend until th application is restarted <-- This is the main reason we still get the "old" plugin version.

Since it was a very deliberate choice, I'm not sure we can simply change the behavior. Could you maybe open an issue on the Theia side so we can open up the discussion to a broader audience? I think the best option for now would be to make a clear distinction between deployed plugins (i.e., multiple versions possible) and active plugins (only the first version is available to the frontend until a reload happens).

**Reply by martin-fleck-at** [reply]
As for a local workaround in case you are blocked, I'd suggest to extend `HostedPluginServerImpl` with a method that does not do the `isRelevantPlugin` check so you can get all plugins in all versions.
